server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java

```diff
@@ -0,0 +1,205 @@
+            if (normalNodes.status == frozenNodes.status) {
+                symptomBuilder.append(NORMAL_GROUP).append(" and ").append(FROZEN_GROUP);
+                diagnoses = List.of(SHARD_LIMITS_REACHED_NORMAL_NODES, SHARD_LIMITS_REACHED_FROZEN_NODES);
+
+            } else if (normalNodes.status.indicatesHealthProblem()) {
+                symptomBuilder.append(NORMAL_GROUP);
+                diagnoses = List.of(SHARD_LIMITS_REACHED_NORMAL_NODES);
+
+            } else if (frozenNodes.status.indicatesHealthProblem()) {
+                symptomBuilder.append(FROZEN_GROUP);
+                diagnoses = List.of(SHARD_LIMITS_REACHED_FROZEN_NODES);
+            }
+
+            symptomBuilder.append(" nodes.");
+        }
+
+        var indicatorImpacts = switch (finalStatus) {
+            case RED -> RED_INDICATOR_IMPACTS;
+            case YELLOW -> YELLOW_INDICATOR_IMPACTS;
+            default -> List.<HealthIndicatorImpact>of();
+        };
+
+        return createIndicator(
+            finalStatus,
+            symptomBuilder.toString(),
+            buildDetails(normalNodes.result, frozenNodes.result),
+            indicatorImpacts,
+            diagnoses
+        );
+    }
+
+    private StatusResult calculateForNormalNodes(HealthMetadata.ShardLimits shardLimits) {
+        int maxShardsPerNodeSetting = shardLimits.maxShardsPerNode();
+        var result = ShardLimitValidator.checkShardLimitForNormalNodes(maxShardsPerNodeSetting, 5, 1, clusterService.state());
+
+        if (result.canAddShards() == false) {
+            return new StatusResult(HealthStatus.RED, result);
+        }
+
+        result = ShardLimitValidator.checkShardLimitForNormalNodes(maxShardsPerNodeSetting, 10, 1, clusterService.state());
+        if (result.canAddShards() == false) {
+            return new StatusResult(HealthStatus.YELLOW, result);
+        }
+
+        return new StatusResult(HealthStatus.GREEN, result);
+    }
+
+    private StatusResult calculateForFrozenNodes(HealthMetadata.ShardLimits shardLimits) {
+        int maxShardsPerNodeSetting = shardLimits.maxShardsPerNodeFrozen();
+        var result = ShardLimitValidator.checkShardLimitForFrozenNodes(maxShardsPerNodeSetting, 5, 1, clusterService.state());
+        if (result.canAddShards() == false) {
+            return new StatusResult(HealthStatus.RED, result);
+        }
+
+        result = ShardLimitValidator.checkShardLimitForFrozenNodes(maxShardsPerNodeSetting, 10, 1, clusterService.state());
+        if (result.canAddShards() == false) {
+            return new StatusResult(HealthStatus.YELLOW, result);
+        }
+
+        return new StatusResult(HealthStatus.GREEN, result);
+    }
+
+    static HealthIndicatorDetails buildDetails(ShardLimitValidator.Result normalNodes, ShardLimitValidator.Result frozenNodes) {
+        return (builder, params) -> {
+            builder.startObject();
+            {
+                builder.startObject(NORMAL_GROUP + "_nodes");
```

Comment:
two comments:

1. it might be worth not splitting the details into groups and just using the same format that the settings `cluster.max_shards_per_node & cluster.max_shards_per_node.frozen`, so the details would be like this:
```json
{
  "max_shards_in_cluster": 18,
  "current_used_shards": 10
  "max_shards_in_cluster.frozen": 1,
  "current_used_shards.frozen": 0
}
```

2. wherever we use `normal & frozen` use non-frozen and frozen. Not sure about this, but I can imagine better wording here.